microceph/ceph/osd.go

```diff
@@ -438,28 +439,407 @@ func AddOSD(s *state.State, path string, wipe bool, encrypt bool) error {
+	}
+	logger.Infof("osd.%d purged", osd)
+	return nil
+}
 
-	// Get the OSDs from the database.
-	err := s.Database.Transaction(s.Context, func(ctx context.Context, tx *sql.Tx) error {
-		records, err := database.GetDisks(ctx, tx)
+func wipeDevice(s common.StateInterface, osd int64) {
+	var err error
+	// get the device path
+	path, _ := database.OSDQuery.Path(s.ClusterState(), osd)
+	// wipe the device, retry with exponential backoff
+	retries := 8
+	var backoff time.Duration
+	for i := 0; i < retries; i++ {
+		err = timeoutWipe(path)
+		if err == nil {
+			// Success: break the retry loop
+			break
+		}
+		// wipe failed - retry after a delay, and make delay exponential
+		logger.Infof("Wipe failed %v, retrying in %v", err, backoff)
+		backoff = time.Duration(math.Pow(2, float64(i))) * time.Millisecond * 100
+		time.Sleep(backoff)
+	}
+	if err != nil {
+		// log a warning, but don't treat wipe failure as a fatal error
+		// e.g. if the device is broken, we still want to remove it from the cluster
+		logger.Warnf("Fault during device wipe: %v", err)
+	}
+}
+
+// timeoutWipe wipes the given device with a timeout, in order not to hang on broken disks
+func timeoutWipe(path string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	_, err := processExec.RunCommandContext(
+		ctx,
+		"dd", "if=/dev/zero",
+		fmt.Sprintf("of=%s", path),
+		"bs=4M", "count=10", "status=none",
+	)
+	return err
+}
+
+func doRemoveOSD(ctx context.Context, s common.StateInterface, osd int64, bypassSafety bool) error {
+	var err error
+
+	// general sanity
+	err = sanityCheck(s, osd)
+	if err != nil {
+		return err
+	}
+
+	if !bypassSafety {
+		// check: at least 3 OSDs
+		err = checkMinOSDs(s, osd)
 		if err != nil {
-			return fmt.Errorf("Failed to fetch disks: %w", err)
+			return err
+		}
+	}
+
+	err = scaleDownFailureDomain(s, osd)
+	if err != nil {
+		return err
+	}
+
+	// check if the osd is still in the cluster -- if we're being re-run, it might not be
+	isPresent, err := haveOSDInCeph(osd)
+	if err != nil {
+		return fmt.Errorf("Failed to check if osd.%d is present in Ceph: %w", osd, err)
+	}
+	// reweight/drain data
+	if isPresent {
+		reweightOSD(ctx, osd, 0)
+	}
+	// perform safety check for stopping
+	if isPresent && !bypassSafety {
+		err = safetyCheckStop(osd)
+		if err != nil {
+			return err
+		}
+	}
+	// take the OSD out and down
+	if isPresent {
+		err = outDownOSD(osd)
+		if err != nil {
+			return err
+		}
+	}
+	// stop the OSD service
+	if isPresent {
+		err = killOSD(osd)
+	}
+	if err != nil {
+		return err
+	}
+	// perform safety check for destroying
+	if isPresent && !bypassSafety {
+		err = safetyCheckDestroy(osd)
+		if err != nil {
+			return err
+		}
+	}
+	// purge the OSD
+	if isPresent {
+		err = purgeOSD(osd)
+		if err != nil {
+			return err
+		}
+	}
+	// Wipe the underlying blocking device
+	wipeDevice(s, osd)
+	// Remove osd config
+	err = removeOSDConfig(osd)
```

Comment:
is this going to succeed if the OSD isn't present?

Reply:
removeOSDConfig() removes the local disk config. It will succeed even if the OSD has been removed from the cluster previously i.e. if the OSD is gone from the cluster as well as if the config dir is ENOENT